### PR TITLE
Support shuffle on Hive partition columns before write

### PR DESCRIPTION
**Pull request by wenleix**
Previously, writing worker will receive rows in all partitions,
and thus can write upper to hive.max-partitions-per-writers partitions.

This session property allows shuffle on partitioned columns when writing
to partitioned unbucketed Hive tables. As a result, rows in the same
partition will be sent to the same writing worker. This increase the
number of maximum partitions written in single query by a factor of
number of total writing workers.

```
== RELEASE NOTES ==

Hive Changes
* Allow shuffle on partitioned columns when writing to partitioned unbucketed Hive tables. This increase the number of maximum partitions written in single query by a factor of number of total writing workers.This behavior has to be explicitly enabled by Connector session property `shuffle_partitioned_columns_for_table_write`. #14010
```

### CODE REVIEW

**Comment by wenleix** [comment]
Supersedes https://github.com/prestodb/presto/pull/13969

**Comment by wenleix** [comment]
cc @mbasmanova , @kaikalur , @aweisberg

**Comment by mbasmanova** [comment]
CC: @biswapesh

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 1]
```diff
@@ -2405,6 +2411,39 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
...
+                .orElseThrow(() -> new TableNotFoundException(tableName));
+
+        Optional<HiveBucketHandle> hiveBucketHandle = getHiveBucketHandle(table);
+        if (hiveBucketHandle.isPresent()) {
+            // For bucketed table, table partitioning (i.e. the bucketing scheme) should be respected,
+            // and there is no additional preferred shuffle partitioning
+            return Optional.empty();
+        }
+
+        if (!(isShufflePartitionedColumnsForTableWriteEnabled(session) || table.getPartitionColumns().isEmpty())) {
```
[the remark this answers is not in the document]
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @arhimondr : Nice catch!

**Comment by wenleix** [comment]
@highker : Wondering if you can take a look at the SPI change?

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java [hunk 2]
```diff
@@ -363,6 +363,11 @@ default void dropColumn(ConnectorSession session, ConnectorTableHandle tableHand
         return Optional.empty();
     }
 
+    default Optional<ConnectorNewTableLayout> getPreferredShuffleLayoutForNewTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: Can we add javadoc to the interface?

File: presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java [hunk 3]
```diff
@@ -393,6 +398,11 @@ default void dropColumn(ConnectorSession session, ConnectorTableHandle tableHand
         return Optional.of(new ConnectorNewTableLayout(partitioningHandle, partitionColumns));
     }
```
Comment (highker) [inline presto-spi/src/main/java/com/facebook/presto/spi/connector/ConnectorMetadata.java]: same here

**Comment by wenleix** [comment]
Thanks @highker  for the review. Comments addressed 😃

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -2418,6 +2425,39 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
...
+            // For bucketed table, table partitioning (i.e. the bucketing scheme) should be respected,
+            // and there is no additional preferred shuffle partitioning
+            return Optional.empty();
+        }
+
+        if (!isShufflePartitionedColumnsForTableWriteEnabled(session) || table.getPartitionColumns().isEmpty()) {
+            return Optional.empty();
+        }
+
+        // TODO: the shuffle partitioning doesn't have to be the same as Hive bucket partitioning
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @wenleix I don't understand this TODO. It seems to me that bucketed tables are handled above and would never reach that code. Would you clarify?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 5]
```diff
@@ -2444,6 +2484,41 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
...
+        }
+
+        List<HiveColumnHandle> columnHandles = getColumnHandles(tableMetadata, ImmutableSet.copyOf(partitionedBy), typeTranslator);
+        Map<String, HiveColumnHandle> columnHandlesByName = Maps.uniqueIndex(columnHandles, HiveColumnHandle::getName);
+        List<Column> partitionColumns = partitionedBy.stream()
+                .map(columnHandlesByName::get)
+                .map(column -> new Column(column.getName(), column.getHiveType(), column.getComment()))
+                .collect(toList());
+
+        // TODO: the shuffle partitioning doesn't have to be the same as Hive bucket partitioning
```
Comment (mbasmanova) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @wenleix Same question about this TODO

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java (hunk 4 above)
Comment (wenleix) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: @mbasmanova : Sorry for the confusion, I mean we don't have to use `HivePartitionHandle` for the shuffle partitioning. Thinking about implementing a different `HiveShufflePartitioningHandle` that distribute the keys more uniformly (Hive bucket function is not that great :) )

**Comment by wenleix** [comment]
@mbasmanova : Thanks @mbasmanova  for the review!

>  With shuffle_partitioned_columns_for_table_write_enabled=true, the data will be shuffles on partition columns and TableWriter operator will run on as many nodes are there are available in the cluster, each node processing distinct set of partitions. E.g. all data for a single partition will be written by the same node

Correct.

>  Within the node, there will be at least 4 threads

It's configurable via `task.writer-count`.  

> Hence, this properly allows to write up to #-nodes x 100 partitions in a single query and avoid making very small files. 

That's accurate. Although I am only thinking write to at most a few thousands partitions in practice 😃

**Comment by mbasmanova** [comment]
@wenleix Thanks for explaining. Sounds great! Can't wait to give it a try.

**Pull request merged**